### PR TITLE
Refactored the ViewController

### PR DESCRIPTION
**Pull request by bdunogier**
> Story: http://jira.ez.no/browse/EZP-24869
> Status: work in progress
> Based on the branch from #1425

This contains the changes to the ViewController for Platform 1.0, unlike the changes from #1425 that might be backported.

It reworks the ViewController and its attached mechanisms: `ViewProvider`, `View`, `Matcher`.

### Major changes
Rendering of (Repository) values is now organized around `View` objects. There is one type of View object per render value type: `ContentView`, `BlockView`... A `ContentView` is an instance of the `View` interface, as well as of `LocationValueView`, that has a `getLocation()` method.

`Views` are created by `ViewBuilder` services. Those are used from the `ViewControllerListener` to build the `View` object based on the Value. The builder can set the default controller, add items to the parameters...

Custom controller actions can get the View object by typehinting it.
- Extra parameters can be added to the view: `$view->addParameters(['relatedContent' => $relatedContent])`
- The template can be changed using `setTemplateIdentifier()`
- The content or location objects that are being rendered can be set (`setLocation()`, `setContent()`...)
- A `Response` object can also be added to the view in order to set custom headers (`setResponse()`) (not implemented yet)

The View is finally returned by the action, and rendered by a Response listener (not implemented yet).

### Backward compatibility
Existing code is supposed to work.

### Internal changes
- The `ControllerManager` is gone
- The `ViewControllerListener` becomes unique.
  It uses `ViewBuilder` objects (`ContentViewBuilder`, etc) to build the view object from the request's attributes, the `ViewConfigurator` to configure the view based on the configuration rules & providers.

### TODO
- [ ] Rebase following the merge of the deprecation to master
- [ ] Implement `setResponse()` on the View interface/object(s).
- [ ] Render `View` objects to `Response` (currently done by the `viewManager`)
- [ ] Update `View\Manager` tests
- [ ] Add tests for `ContentViewBuilder` and `BlockViewBuilder`
- [ ] Add required packages (`ezsystems/comments-bundle`, others ?)
- [x] Update `ViewControllerListenerTest`
- [x] BC: Deprecate `(Block|Content|Location)ViewPass`.  Change to register custom ViewProviders (untag the `Configured` ones) only, with a deprecation message.

### CODE REVIEW

File: eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php [hunk 1]
```diff
@@ -76,42 +94,51 @@ public function getController(FilterControllerEvent $event)
...
+        if ($request->attributes->has('contentId')) {
+            $contentId = $request->attributes->get('contentId');
+            $content = $this->repository->sudo(
+                function (Repository $repository) use ($contentId) {
+                    return $repository->getContentService()->loadContent($contentId);
+                }
+            );
 
-            return;
+            $location = ($request->attributes->has('location')) ? ($request->attributes->get('location')) : null;
```
Comment (lolautruche) [inline eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php]: Can be simplified to:

```php
$location = $request->attributes->get('location');
```

`get()` will return `null` if the attribute is not defined. If you want it to return a default value, just specify it as a 2nd argument.

File: eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php [hunk 2]
```diff
@@ -76,42 +94,51 @@ public function getController(FilterControllerEvent $event)
...
+
+            if ($view->viewType == 'embed' && !$this->canRead($content, $location)) {
+                throw new UnauthorizedException('content', 'read|view_embed', ['contentId' => $contentId]);
+            }
+
+            $request->attributes->set('content', $content);
+            $view->setContent($content);
+        } elseif ($request->attributes->has('content') && $request->attributes->get('content') instanceof Content) {
```
Comment (lolautruche) [inline eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php]:
```php
if (// ...) {
    // ...
} elseif ($valueObject = $request->attributes->get('content') instanceof Content) {
    // ...
}
```

Comment (lolautruche) [inline eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php]: And then you don't need the line just below

File: eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php (hunk 1 above)
Comment (bdunogier) [inline eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php]: Cool :)

File: eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php (hunk 2 above)
Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php]: won't that make the `$valueObject` a bool? better add parenthesis imho for readability :)

**Comment by andrerom** [comment]
I'm not that familiar with this, but as long as you and @lolautruche agree on the direction here, then it so far looks good to me.

**Comment by bdunogier** [comment]
Even though it doesn't show here, most tests are passing on the BDD demo scenario, at least :-)

**Comment by andrerom** [comment]
BTW: Most likely reason why this is not tested on travis is that it is pointing to a branch not configured to be tested `ezp24869-viewContent_deprecation`.

**Comment by bdunogier** [comment]
Hmm, odd. I've done this kind of nested prs before, and it was executed.
I'll try anyway, thank you :)​

**Comment by bdunogier** [comment]
You were right it seems @andrerom. (Added)[https://github.com/ezsystems/ezpublish-kernel/commit/496a31a769ded616cc561b12a5d1c1bdef33a316] the parent branch name to `.travis.yml`, and it's testing now.

Let the failures fly... :-)

**Comment by andrerom** [comment]
:)

**Comment by emodric** [comment]
Just a note not to forget changes done in #1444 / #1447  :)

**Comment by bdunogier** [comment]
Will reopen against master.

**Pull request closed without merging**